neonvm/controllers/vm_controller_test.go

```diff
@@ -1,136 +1,342 @@
-/*
-Copyright 2022.
-
-Licensed under the Apache License, Version 2.0 (the "License");
-you may not use this file except in compliance with the License.
-You may obtain a copy of the License at
-
-    http://www.apache.org/licenses/LICENSE-2.0
-
-Unless required by applicable law or agreed to in writing, software
-distributed under the License is distributed on an "AS IS" BASIS,
-WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-See the License for the specific language governing permissions and
-limitations under the License.
-*/
-
 package controllers
 
 import (
 	"context"
+	"encoding/json"
 	"os"
-	"time"
+	"testing"
 
-	. "github.com/onsi/ginkgo/v2"
-	. "github.com/onsi/gomega"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap/zapcore"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
 	corev1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/api/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/types"
 
 	vmv1 "github.com/neondatabase/autoscaling/neonvm/apis/neonvm/v1"
+	"github.com/neondatabase/autoscaling/pkg/util"
 )
 
-var _ = Describe("VirtualMachine controller", func() {
-	Context("VirtualMachine controller test", func() {
+type clientMock struct {
+	objects map[types.NamespacedName][]byte
+	t       *testing.T
+}
```

Comment:
> Do you think there are arguments against manual reimplementation?

I do, yeah. First, the K8s API is quite complicated. For example, neonvm-controller uses JSON patch to update pod metadata, but I see that's not implemented here. You _could_ reimplement all that, but why bother?

Second: I think it's better to use the mock generated by the k8s related tools rather than writing our own, otherwise we risk mocking based on our own (potentially flawed) understanding, increasing the risk that our tests are blind to behaviors we didn't expect from k8s.

To your point about it making it easier to simulate failures... AFAICT it should be possible by adding a `Reactor` to the [`ReactionChain`](https://github.com/kubernetes/client-go/blob/94cabc85d99585c7437d1b80f9c2571e5e6cf879/testing/fake.go#L38) in `Fake` to allow externally injected failures. (could use e.g. `PrependReactor` but then it's restricted to a single resource + action)

Reply:
Ok, good points. Let's merge it like this, and I'll fix it in the follow-up?

Reply:
Actually fixed it right here, please, take a look.

Reply:
nice!